tests/test_translation.py

```diff
@@ -108,57 +108,54 @@ def test_invest_validate(self):
-                    cli.main(
-                        ['--language', TEST_LANG, 'validate', datastack_path])
+        from natcap.invest import cli
+        with patch('sys.stdout', new=io.StringIO()) as out:
+            with self.assertRaises(SystemExit):
+                cli.main(
+                    ['--language', TEST_LANG, 'validate', datastack_path])
 
         result = out.getvalue()
         self.assertIn(TEST_MESSAGES[missing_key_msg], result)
 
     def test_server_get_invest_models(self):
         """Translation: test that /models endpoint is translated."""
-        with patch('natcap.invest.LOCALE_DIR', self.test_locale_dir):
-            test_client = ui_server.app.test_client()
-            response = test_client.get('api/models')
-            result = json.loads(response.get_data(as_text=True))
+        test_client = ui_server.app.test_client()
+        response = test_client.get(
+            'api/models', query_string={'language': TEST_LANG})
+        result = json.loads(response.get_data(as_text=True))
         self.assertIn(
             TEST_MESSAGES['Carbon Storage and Sequestration'], result)
 
     def test_server_get_invest_getspec(self):
         """Translation: test that /getspec endpoint is translated."""
-        with patch('natcap.invest.LOCALE_DIR', self.test_locale_dir):
-            test_client = ui_server.app.test_client()
-            response = test_client.post('api/getspec', json='carbon')
-            spec = json.loads(response.get_data(as_text=True))
+        test_client = ui_server.app.test_client()
+        response = test_client.post(
+            'api/getspec', json='carbon', query_string={'language': TEST_LANG})
+        spec = json.loads(response.get_data(as_text=True))
         self.assertEqual(
             spec['args']['lulc_cur_path']['name'],
             TEST_MESSAGES['current LULC'])
 
     def test_server_get_invest_validate(self):
         """Translation: test that /validate endpoint is translated."""
-        with patch('natcap.invest.LOCALE_DIR', self.test_locale_dir):
-            test_client = ui_server.app.test_client()
-            payload = {
-                'model_module': carbon.ARGS_SPEC['pyname'],
-                'args': json.dumps({})
-            }
-            response = test_client.post('api/validate', json=payload)
+        test_client = ui_server.app.test_client()
+        payload = {
+            'model_module': carbon.ARGS_SPEC['pyname'],
+            'args': json.dumps({})
+        }
+        response = test_client.post(
+            'api/validate', json=payload,
+            query_string={'language': TEST_LANG})
         results = json.loads(response.get_data(as_text=True))
         messages = [item[1] for item in results]
         self.assertIn(TEST_MESSAGES[missing_key_msg], messages)
 
     def test_translate_formatted_string(self):
-        with patch('natcap.invest.LOCALE_DIR', self.test_locale_dir):
-            install_language('en')
+        set_locale('ll')
```

Comment:
It looks like `TEST_LANGUAGE` is 'll', so could use that variable? But it's curious that the language is the only change to this test and it still passes? This one also looks a lot like the previous test, are they both needed?

Reply:
I changed the test language for all tests from "en" to "ll" (the foo of language codes) to make it more explicit that we're using a fake, test language.

The first three tests are for translation through the CLI for the 3 commands where it's relevant: `list`, `getspec`, `validate`.
The next three tests are for translation through the 3 corresponding server endpoints: `api/models`, `api/getspec`, `api/validate`.
This last one tests translation more directly, not through either interface. That's why it's the only one that calls `set_locale` directly. The CLI or server endpoint calls `set_locale` in the other tests. Its purpose is testing the translation functionality independent of the CLI or server.
It also tests that translation works on an f-string, which is important because formatting an f-string triggers evaluation of the string, which calls the translation function.